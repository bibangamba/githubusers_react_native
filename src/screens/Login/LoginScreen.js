import React, { Component } from "react";
import {
  View,
  Text,
  Image,
  TextInput,
  TouchableOpacity,
  KeyboardAvoidingView,
  AsyncStorage,
  Platform
} from "react-native";
import githubLogin from "../../utils/githubLogin";
import styles from "./LoginScreen.style";

class LoginScreen extends Component {
  constructor(props) {
    super(props);
  }

  state = {
    username: "",
    password: "",
    error: ""
  };

  onUsernameFieldChange = username => {
    this.setState({ username: username });
  };

  onPasswordFieldChange = password => {
    this.setState({ password: password });
  };

  onSubmit = () => {
    const { username, password } = this.state;
    githubLogin(username, password).then(result => {
      console.log("RESULT from auth fetch: ", result);

      if (typeof result === "object" && result !== null) {
        console.log("############ got an error: ", result["error"]);
      }

      AsyncStorage.setItem("access_token", result)
        .then(() => console.log("// saved data to async storage"))
        .catch(error => console.log("// Error saving data ", error));
      const { navigation } = this.props;
      navigation.navigate("UsersList");
    });
  };

  render() {
    const verticalKeyboardOffset = 500;
    return (
      <KeyboardAvoidingView
        style={styles.viewContainer}
        behavior="padding"
        keyboardVerticalOffset={Platform.select({
          ios: 0,
          android: verticalKeyboardOffset
        })}
      >
        <Image
          style={styles.githubLogo}
          source={require("../../../assets/github.png")}
        />
        <View style={styles.fieldContainer}>
          <TextInput
            onChangeText={this.onUsernameFieldChange}
            style={styles.textInput}
            placeholder="username"
            autoCapitalize="none"
            textContentType="username"
            placeholderTextColor="#888888"
          />
          <TextInput
            onChangeText={this.onPasswordFieldChange}
            style={styles.textInput}
            placeholder="password"
            autoCapitalize="none"
            placeholderTextColor="#888888"
            textContentType="password"
            secureTextEntry={true}
          />
          <TouchableOpacity style={styles.signinButton} onPress={this.onSubmit}>
            <Text style={styles.signinText}>Sign in</Text>
          </TouchableOpacity>
        </View>
      </KeyboardAvoidingView>
    );
  }
}

export default LoginScreen;
