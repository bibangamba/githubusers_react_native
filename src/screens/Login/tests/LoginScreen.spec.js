import { shallow, mount, render } from "enzyme";
import LoginScreen from "../LoginScreen";

import React from "react";
import renderer from "react-test-renderer";
const mock_password = "yolo-bibz#007";
const mock_username = "bibz";
const fetch = jest.fn(() => new Promise(resolve => resolve()));
const mockSuccesfulResponse = (status = 201, method = "POST") => {
  global.fetch = jest.fn().mockImplementationOnce(() => {
    return new Promise((resolve, reject) => {
      resolve({
        ok: true,
        status,
        json: () => {
          return returnBody ? returnBody : {};
        }
      });
    });
  });
};
describe("Test LoginScreen", () => {
  it("should match to snapshot", () => {
    const component = shallow(<LoginScreen />);
    expect(component).toMatchSnapshot();
  });

  it("should render the login screen without crashing", () => {
    const renderedScreen = renderer.create(<LoginScreen />).toJSON();
    expect(renderedScreen).toBeTruthy();
  });

  it("should render the header image", () => {
    const loginScreen = shallow(<LoginScreen />);
    expect(loginScreen.find("Image")).toHaveLength(1);
    expect(loginScreen.find("TextInput")).toHaveLength(2);
  });

  it("should change state if username is entered", () => {
    const instanceOfLoginScreen = renderer
      .create(<LoginScreen />)
      .getInstance();
    instanceOfLoginScreen.onUsernameFieldChange(mock_username);
    instanceOfLoginScreen.onPasswordFieldChange(mock_password);
    mockSuccesfulResponse();
    instanceOfLoginScreen.onSubmit();
    expect(instanceOfLoginScreen.state.username).toEqual(mock_username);
    expect(instanceOfLoginScreen.state.password).toEqual(mock_password);
  });

  //   it("should change state when text input values change", () => {
  //     const loginScreen = mount(<LoginScreen />);
  //     console.log(loginScreen);
  //     const usernameInput = loginScreen.find(input).at(0);
  //     const passwordInput = loginScreen.find(input).at(1);
  //     usernameInput.instance().value = mock_username;
  //     usernameInput.simulate("change");
  //     passwordInput.instance().value = mock_password;
  //     passwordInput.simulate("change");
  //     expect(loginScreen.state().username).toEqual(mock_username);
  //     expect(loginScreen.state().password).toEqual(mock_password);
  //   });
});
