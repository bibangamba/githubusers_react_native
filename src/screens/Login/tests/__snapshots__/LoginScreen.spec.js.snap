// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LoginScreen should match to snapshot 1`] = `ShallowWrapper {}`;
